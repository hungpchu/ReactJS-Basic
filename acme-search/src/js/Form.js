import React from "react";
import ".././styles/Form.css";

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dataSourceValue: "select",
      dropdownValue: ["select", "contacts", "dropbox", "slack", "tweet"],
      contactForm: [
        "id",
        "name",
        "company",
        "emails",
        "phones",
        "matching_terms",
        "last_contact",
      ],
      slackForm: [
        "id",
        "channel",
        "alice",
        "message",
        "timestamp",
        "matching_terms",
      ],
      tweetForm: ["user", "message", "timestamp", "matching_terms"],
      dropboxForm: [
        "id",
        "path",
        "title",
        "shared_with",
        "matching_terms",
        "created",
      ],
      array: ["emails", "phones", "shared_with", "matching_terms"],
      newObject: {},
    };
    this.handleSelectionChange = this.handleSelectionChange.bind(this);
    this.displayForm = this.displayForm.bind(this);
    this.updateNewObject = this.updateNewObject.bind(this);
    this.saveData = this.saveData.bind(this);
    this.displayDropDownValue = this.displayDropDownValue.bind(this);
  }

  handleSelectionChange(event) {
    this.setState({ dataSourceValue: event.target.value });
    this.setState({ newObject: {} });
  }

  updateNewObject(key, event) {
    const isArray = this.state.array.includes(key.element);
    const updatedNewObject = this.state.newObject;
    isArray
      ? (updatedNewObject[key.element] = event.target.value.split(","))
      : (updatedNewObject[key.element] = event.target.value);
    this.setState({ newObject: updatedNewObject });
  }

  displayForm() {
    if (this.state.dataSourceValue === "select") return null;
    const array =
      this.state.dataSourceValue === "contacts"
        ? this.state.contactForm
        : this.state.dataSourceValue === "dropbox"
        ? this.state.dropboxForm
        : this.state.dataSourceValue === "slack"
        ? this.state.slackForm
        : this.state.tweetForm;
    const field = array.map((element) => (
      <li key={JSON.stringify(element)} id={element + "-field-item"}>
        <label id={element + "-label"}>{element}</label>
        <input
          type="text"
          id={element + "-input"}
          placeholder={element}
          onChange={this.updateNewObject.bind(this, { element })}
        />
        {this.state.array.includes(element) && (
          <p className="instruction-message">
            *{element} is array. Please add ',' in between for multiple value
          </p>
        )}
      </li>
    ));
    return (
      <>
        <ul className="field-input-list">{field}</ul>
        <button id="save-data-button" onClick={this.saveData}>
          Save data
        </button>
      </>
    );
  }

  saveData() {
    this.props.saveUpdatedData(
      this.state.dataSourceValue,
      this.state.newObject
    );
    // set the newObject back to empty
    // for the new updated one later
    this.setState({ newObject: {} });
  }

  displayDropDownValue() {
    return this.state.dropdownValue.map((value) => (
      <option
        id={value + "-selection-value"}
        type="submit"
        key={value + "-selection-value"}
      >
        {value}
      </option>
    ));
  }

  render() {
    return (
      <div className="update-form-body">
        <label className="dropdown-label">
          Choose a datasource to save updated data:
        </label>
        <select
          className="dropdown-selection"
          onClick={this.handleSelectionChange}
        >
          {this.displayDropDownValue()}
        </select>
        {this.displayForm()}
      </div>
    );
  }
}

export default Form;
