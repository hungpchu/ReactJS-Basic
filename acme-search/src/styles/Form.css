select {
  font-size: 0.9rem;
  padding: 2px 5px;
}

ul > li {
  list-style-type: none;
}

/* for the label in form */
li > label {
  display: inline-block;
  color: white;
  width: 40px;
  margin-bottom: 20px;
  margin-left: -40px;
}

.instruction-message {
  color: red;
  font-size: 0.8rem;
  margin-top: -15px;
  margin-left: 100px;
}

.update-form-body {
  /* Center the form on the page */
  width: 600px;
  /* Form outline */
  padding: 1em;
  border: 1px solid white;
  border-radius: 1em;
  margin: 0 auto;
}

.update-form-body > label {
  color: whitesmoke;
  font-size: 1.2rem;
  padding-right: 10px;
}

#save-data-button {
  margin-left: 500px;
}
