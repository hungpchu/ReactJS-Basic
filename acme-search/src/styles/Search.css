ol {
  margin: 30px 30px 30px 100px;
  color: whitesmoke;
}
ol > li {
  margin-bottom: 20px;
}

.seach-body {
  align-items: center;
  justify-content: center;
}

.seach-box {
  display: block;
  margin-left: auto;
  margin-right: auto;
  text-align: center;
  margin-bottom: 30px;
  margin-right: 80px;
}

#update-data-button {
  margin-left: 15px;
}
