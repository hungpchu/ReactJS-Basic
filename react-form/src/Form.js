import React from 'react';


class Form extends React.Component{
    constructor(props)
    {
        super(props);
        this.state = 
        { 
            text: "",
            age:  "" 
        };
    }

   handleChangeName = (event) => {
       this.setState({text : event.target.value});
   }

   handleChangeAge = (event) => {
    this.setState({age: event.target.value});
}

   handleSubmit = ( event ) => {
       alert("Value of submission text : " + this.state.text + " Value of submission age: " + this.state.age);
       event.preventDefault();
   }

   render() {

    return(
        
        <form onSubmit={this.handleSubmit}>
            <label>
                Name: <input type="text" value={this.state.text} onChange={this.handleChangeName} />
                </label>
                <br/>
                <label>
                Age: <input type="text" value={this.state.age} onChange={this.handleChangeAge} />
            </label>
            <br/>
            <input type="submit" value="Submit" />
        </form>
    );
   }
}

export default Form;