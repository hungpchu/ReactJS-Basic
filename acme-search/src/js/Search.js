import React from "react";
import calendarData from ".././data/calendar.json";
import contactData from ".././data/contacts.json";
import dropboxData from "../data/dropbox.json";
import slackData from "../data/slack.json";
import tweetData from "../data/tweet.json";
import Form from "./Form.js";
import ".././styles/Search.css";

/***
 * Search component:
 * 1. deal with input from user as searchKey and return result
 * 2. show updated data form when user clicks Update data
 */
class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchKey: "",
      visible: false,
      calendar: calendarData.calendar,
      contact: contactData.contacts,
      dropbox: dropboxData.dropbox,
      slack: slackData.slack,
      tweet: tweetData.tweet,
      resultList: [],
      timeArray: ["date", "last_contact", "created", "timestamp"],
    };

    this.setSearchKeyChange = this.setSearchKeyChange.bind(this);
    this.search = this.search.bind(this);
    this.visible = this.visible.bind(this);
    this.saveUpdatedData = this.saveUpdatedData.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
  }

  /***
   * function to format array for display
   */
  formatArray(array) {
    return array?.map((value, index, array) =>
      index !== array.length - 1 ? value + ", " : value
    );
  }

  /***
   * function to format time for display
   */
  formatTime(time) {
    time = time.split(" ")[0].split("-");
    return time[1] + "/" + time[2] + "/" + time[0];
  }

  /***
   * function to update the value of searchKey
   * when user type in the new search Key
   * */
  setSearchKeyChange(event) {
    this.setState({ searchKey: event.target.value });
  }

  /***
   * function to search while pressing key
   * without click search
   */
  onKeyPress(event) {
    var code = event.keyCode || event.which;
    // code = 13 is for Enter key
    if (code === 13) this.search();
  }

  /***
   * function to search the value of searchKey
   * in the dataSource
   */
  search() {
    // new updated result List
    let updatedResultList = [];
    updatedResultList = updatedResultList.concat(
      this.state.calendar.filter((element) =>
        element.matching_terms?.some((key) =>
          this.state.searchKey.includes(key)
        )
      )
    );
    updatedResultList = updatedResultList.concat(
      this.state.contact.filter((element) =>
        element.matching_terms?.some((key) =>
          this.state.searchKey.includes(key)
        )
      )
    );
    updatedResultList = updatedResultList.concat(
      this.state.dropbox.filter((element) =>
        element.matching_terms?.some((key) =>
          this.state.searchKey.includes(key)
        )
      )
    );
    updatedResultList = updatedResultList.concat(
      this.state.slack.filter((element) =>
        element.matching_terms?.some((key) =>
          this.state.searchKey.includes(key)
        )
      )
    );
    updatedResultList = updatedResultList.concat(
      this.state.tweet.filter((element) =>
        element.matching_terms?.some((key) =>
          this.state.searchKey.includes(key)
        )
      )
    );
    this.setState({ resultList: updatedResultList });
  }

  /***
   * function to update the visible of
   * updated data form
   */
  visible() {
    this.setState({ visible: !this.state.visible });
  }

  /***
   * function to pass on to child component
   * Form to save updated Data input by user
   */
  saveUpdatedData(key, newObj) {
    if (key === "select" || Object.keys(newObj).length === 0) return;
    let dataSource;
    if (key === "contacts") {
      dataSource = this.state.contact.concat(newObj);
      this.setState({ contact: dataSource });
    } else if (key === "dropbox") {
      dataSource = this.state.dropbox.concat(newObj);
      this.setState({ dropbox: dataSource });
    } else if (key === "slack") {
      dataSource = this.state.slack.concat(newObj);
      this.setState({ slack: dataSource });
    } else {
      dataSource = this.state.tweet.concat(newObj);
      this.setState({ tweet: dataSource });
    }
    // close the updated data form after done
    this.visible();
  }

  /***
   * function to map result in resultList to display
   * in the screen for the user
   */
  showResult() {
    if (this.state.resultList.length === 0) return null;
    const resultDisplay = this.state.resultList.map((data) => {
      return (
        <li key={JSON.stringify(data)}>
          {Object.keys(data).map((key) => (
            <p
              key={JSON.stringify(key) + JSON.stringify(data[key])}
              id={"value-of-" + key + "-key"}
            >
              {key}:{" "}
              {this.state.timeArray.includes(key)
                ? this.formatTime(data[key])
                : typeof data[key] === "string"
                ? data[key]
                : this.formatArray(data[key])}{" "}
            </p>
          ))}
        </li>
      );
    });
    return <ol className="result-list">{resultDisplay}</ol>;
  }

  render() {
    return (
      <div className="seach-body">
        <div className="seach-box">
          <input
            className="search-input"
            type="text"
            placeholder="Search"
            onChange={this.setSearchKeyChange}
            onKeyPress={this.onKeyPress}
          />
          <button className="search-button" onClick={this.search}>
            Search
          </button>
          <button id="update-data-button" onClick={this.visible}>
            Update Data
          </button>
        </div>
        {this.state.visible && <Form saveUpdatedData={this.saveUpdatedData} />}
        {this.state.searchKey.length > 0 && this.showResult()}
      </div>
    );
  }
}

export default Search;
