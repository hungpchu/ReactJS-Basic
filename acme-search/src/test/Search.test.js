import React from "react";
import renderer from "react-test-renderer";
import Search from "./../js/Search.js";
import Form from "./../js/Form.js";
import Adapter from "enzyme-adapter-react-16";
import { mount, configure } from "enzyme";

configure({ adapter: new Adapter() });
describe("<Search />", () => {
  let mockSearch, mockForm;
  beforeEach(() => {
    mockSearch = mount(<Search />);
    mockForm = mount(<Form />);
  });
  describe("(snapshot)", () => {
    test("renders correctly", () => {
      const searchSnapShot = renderer.create(<Search />).toJSON();
      expect(searchSnapShot).toMatchSnapshot();
    });
  });
  describe("the visibility of the Form", () => {
    test("should not render Form if visible is false", () => {
      mockSearch.setState({
        visible: false,
      });
      expect(mockSearch.find(".update-form-body").exists()).toEqual(false);
    });

    test("should render Form if visible is true", () => {
      mockSearch.setState({
        visible: true,
      });
      expect(mockSearch.find(".update-form-body").exists()).toEqual(true);
    });

    test("should render Form if click the update data button", () => {
      mockSearch.find("#update-data-button").simulate("click");
      expect(mockSearch.find(".update-form-body").exists()).toEqual(true);
    });
  });

  describe("the visibility of the Result list", () => {
    test("should render result list with one search key matches", () => {
      mockSearch.setState({
        searchKey: "other",
      });
      mockSearch.find(".search-button").simulate("click");
      expect(mockSearch.find(".result-list").exists()).toEqual(true);
    });

    test("should render result list with multiple search key matches", () => {
      mockSearch.setState({
        searchKey: "other,alice,acme",
      });
      mockSearch.find(".search-button").simulate("click");
      expect(mockSearch.find(".result-list").exists()).toEqual(true);
    });

    test("should render result-list with one key matches and other not", () => {
      mockSearch.setState({
        searchKey: "other,hungchu",
      });
      mockSearch.find(".search-button").simulate("click");
      expect(mockSearch.find(".result-list").exists()).toEqual(true);
    });

    test("should not render result-list if search key not match", () => {
      mockSearch.setState({
        searchKey: "hungchu",
      });
      mockSearch.find(".search-button").simulate("click");
      expect(mockSearch.find(".result-list").exists()).toEqual(false);
    });
  });
});
