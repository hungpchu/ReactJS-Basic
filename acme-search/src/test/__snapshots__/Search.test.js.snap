// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> (snapshot) renders correctly 1`] = `
<div
  className="seach-body"
>
  <div
    className="seach-box"
  >
    <input
      className="search-input"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Search"
      type="text"
    />
    <button
      className="search-button"
      onClick={[Function]}
    >
      Search
    </button>
    <button
      id="update-data-button"
      onClick={[Function]}
    >
      Update Data
    </button>
  </div>
</div>
`;
